app/controllers/predicates/AuthorisedAction.scala

```diff
@@ -18,64 +18,77 @@ package controllers.predicates
 
-class AuthorisedAction @Inject()(
-                                  appConfig: AppConfig,
-                                  val agentAuthErrorPage: AgentAuthErrorPageView
-                                )(
-                                  implicit val authService: AuthService,
-                                  val mcc: MessagesControllerComponents
-                                ) extends ActionBuilder[User, AnyContent] with I18nSupport{
+class AuthorisedAction @Inject()(appConfig: AppConfig,
+                                 val agentAuthErrorPage: AgentAuthErrorPageView)
+                                (implicit val authService: AuthService,
+                                 val mcc: MessagesControllerComponents) extends ActionBuilder[User, AnyContent] with I18nSupport {
 
   implicit val executionContext: ExecutionContext = mcc.executionContext
   lazy val logger: Logger = Logger.apply(this.getClass)
   implicit val config: AppConfig = appConfig
-  implicit val messagesApi = mcc.messagesApi
+  implicit val messagesApi: MessagesApi = mcc.messagesApi
+
+  val minimumConfidenceLevel: Int = ConfidenceLevel.L200.level
 
   override def parser: BodyParser[AnyContent] = mcc.parsers.default
 
   override def invokeBlock[A](request: Request[A], block: User[A] => Future[Result]): Future[Result] = {
 
     implicit lazy val headerCarrier: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
 
-    authService.authorised(ConfidenceLevel.L200).retrieve(allEnrolments and affinityGroup) {
-      case enrolments ~ Some(AffinityGroup.Agent) =>
+    authService.authorised.retrieve(allEnrolments and affinityGroup and confidenceLevel) {
+      case enrolments ~ Some(AffinityGroup.Agent) ~ _ =>
         checkAuthorisation(block, enrolments, isAgent = true)(request, headerCarrier)
-      case enrolments ~ _ =>
+      case enrolments ~ _ ~ confidenceLevel if confidenceLevel.level >= minimumConfidenceLevel =>
         checkAuthorisation(block, enrolments)(request, headerCarrier)
+      case _ =>
+        logger.info("[AuthorisedAction][invokeBlock] User has confidence level below 200, routing user to IV uplift.")
+        Future(Redirect(routes.IVUpliftController.initialiseJourney()))
     } recover {
       case _: NoActiveSession =>
         logger.info(s"[AgentPredicate][authoriseAsAgent] - No active session. Redirecting to ${appConfig.signInUrl}")
         Redirect(appConfig.signInUrl) //TODO Check this is the correct location
+      case e: NoSuchElementException =>
+        if (e.getMessage.contains("Illegal confidence level")) {
+          logger.error("[AuthorisedAction][invokeBlock] User has invalid confidence level, routing user to IV uplift.")
+          Redirect(routes.IVUpliftController.initialiseJourney())
+        } else {
+          logger.error("[AuthorisedAction][invokeBlock] Received NoSuchElementException form auth when retrieving confidence level.")
```

Comment:
This log is possibly misleading, one of the retrievals that aren't confidenceLevel could be missing resulting in this log.

Reply:
Updated